### ISSUE TITLE
Unknown issue is reducing number of SynGO annotations in funal MGI pipeline products

### ISSUE THREAD

**Issue by kltm** [geneontology/go-site#1847]
Currently, there is some as-yet unknown issue where the number of SynGO annotations for MGI, coming in from Noctua, has been greatly reduced since the last release. From a recent release trial run:

MGI upstream src:
```
sjcarbon@moiraine:/tmp/foo$:) zgrep -i syngo mgi-src.gaf.gz  | wc -l
155
```
MGI raw noctua:
```
sjcarbon@moiraine:/tmp/foo$:) zgrep -i syngo noctua_mgi_valid.gaf.gz  | wc -l
6919
```
MGI final merged:
```
sjcarbon@moiraine:/tmp/foo$:) zgrep -i syngo mgi.gaf.gz  | wc -l
312
```

Tagging @dustine32 @pgaudet

**Comment by kltm** [comment]
One current theory (@dustine32 ) is that something has gone wrong in the merge.

**Comment by pgaudet** [comment]
Thanks. Error reports (http://snapshot.geneontology.org/reports/noctua_mgi-report.html) dont explain the drop; there are only ~300 errors and we lost about 3000 annotations. 

Thanks, Pascale

**Comment by ukemi** [comment]
I don't think it is because of errors. Although we filter A LOT of SynGO annotations because of errors, in MGI production I see 538 MGI markers with SynGO annotations loaded from Noctua. Some of these markers have multiple annotations from SynGO associated with them. 
Here are a few lines from: http://snapshot.geneontology.org/products/annotations/noctua_mgi.gpad

MGI	MGI:88276	is_active_in	GO:0098831	PMID:8909549	ECO:0005593			20180711	SynGO	part_of(UBERON:0002037)	contributor=https://orcid.org/0000-0003-3185-5709|noctua-model-id=gomodel:SYNGO_1000|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:88276	is_active_in	GO:0098831	PMID:8909549	ECO:0006003			20180711	SynGO	part_of(UBERON:0002037)	contributor=https://orcid.org/0000-0003-3185-5709|noctua-model-id=gomodel:SYNGO_1000|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:88276	is_active_in	GO:0098831	PMID:8909549	ECO:0006003			20180711	SynGO	part_of(UBERON:0000956)	contributor=https://orcid.org/0000-0003-3185-5709|noctua-model-id=gomodel:SYNGO_1000|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:88276	is_active_in	GO:0098831	PMID:8909549	ECO:0005593			20180711	SynGO	part_of(UBERON:0000956)	contributor=https://orcid.org/0000-0003-3185-5709|noctua-model-id=gomodel:SYNGO_1000|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:88276	is_active_in	GO:0098685	PMID:14622577	ECO:0001225			20180711	SynGO		contributor=https://orcid.org/0000-0003-3185-5709|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|noctua-model-id=gomodel:SYNGO_1001|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:88276	involved_in	GO:0097091	PMID:14622577	ECO:0006003			20180711	SynGO	occurs_in(GO:0098685)	contributor=https://orcid.org/0000-0003-3185-5709|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|noctua-model-id=gomodel:SYNGO_1001|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:88276	involved_in	GO:0097091	PMID:14622577	ECO:0001225			20180711	SynGO	occurs_in(GO:0098685)	contributor=https://orcid.org/0000-0003-3185-5709|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|noctua-model-id=gomodel:SYNGO_1001|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:88276	is_active_in	GO:0098685	PMID:14622577	ECO:0006003			20180711	SynGO		contributor=https://orcid.org/0000-0003-3185-5709|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|noctua-model-id=gomodel:SYNGO_1001|contributor=https://orcid.org/0000-0002-0755-5899
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0006062			20180711	SynGO	part_of(UBERON:0001950)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0005593			20180711	SynGO	part_of(UBERON:0001950)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0006003			20180711	SynGO	part_of(UBERON:0002435)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0005593			20180711	SynGO	part_of(UBERON:0002435)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0006062			20180711	SynGO	part_of(UBERON:0002435)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0006065			20180711	SynGO	part_of(UBERON:0002435)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0006003			20180711	SynGO	part_of(UBERON:0001950)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0006065			20180711	SynGO	part_of(UBERON:0001950)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
MGI	MGI:108110	involved_in	GO:0099560	PMID:26771491	ECO:0000353			20180711	SynGO	occurs_in(GO:0098978),occurs_in(UBERON:0001897)	contributor=https://orcid.org/0000-0002-0269-0311|contributor=https://orcid.org/0000-0002-9681-3318|contributor=https://orcid.org/0000-0002-9880-1214|model-state=production|contributor=https://orcid.org/0000-0001-9377-7414|noctua-model-id=gomodel:SYNGO_1004
MGI	MGI:108110	involved_in	GO:0099560	PMID:26771491	ECO:0005593			20180711	SynGO	occurs_in(GO:0098978),occurs_in(UBERON:0001897)	contributor=https://orcid.org/0000-0002-0269-0311|contributor=https://orcid.org/0000-0002-9681-3318|contributor=https://orcid.org/0000-0002-9880-1214|model-state=production|contributor=https://orcid.org/0000-0001-9377-7414|noctua-model-id=gomodel:SYNGO_1004
MGI	MGI:108110	involved_in	GO:0099560	PMID:26771491	ECO:0006065			20180711	SynGO	occurs_in(GO:0098978),occurs_in(UBERON:0001897)	contributor=https://orcid.org/0000-0002-9681-3318|contributor=https://orcid.org/0000-0002-0269-0311|contributor=https://orcid.org/0000-0002-9880-1214|model-state=production|contributor=https://orcid.org/0000-0001-9377-7414|noctua-model-id=gomodel:SYNGO_1004
MGI	MGI:108110	is_active_in	GO:0098978	PMID:26771491	ECO:0005589			20180711	SynGO	part_of(UBERON:0001897)	contributor=https://orcid.org/0000-0002-0269-0311|contributor=https://orcid.org/0000-0002-9681-3318|contributor=https://orcid.org/0000-0002-9880-1214|model-state=production|contributor=https://orcid.org/0000-0001-9377-7414|noctua-model-id=gomodel:SYNGO_1004
MGI	MGI:108110	involved_in	GO:0099560	PMID:26771491	ECO:0001225			20180711	SynGO	occurs_in(GO:0098978),occurs_in(UBERON:0001897)	contributor=https://orcid.org/0000-0002-0269-0311|contributor=https://orcid.org/0000-0002-9681-3318|contributor=https://orcid.org/0000-0002-9880-1214|model-state=production|contributor=https://orcid.org/0000-0001-9377-7414|noctua-model-id=gomodel:SYNGO_1004

**Comment by kltm** [comment]
After talking to @dustine32 , it seems like the issue may be related to ECO -> 3-letter code mapping. We believe it may well be another expression of https://github.com/geneontology/pipeline/issues/283 .

@pgaudet We are going to approach this from two angles. The first is an immediate fix that @dustine32 is working on that makes sure that if an ECO id does not map it is reported as an error and that annotation cannot proceed (the problem seems to be that things can proceed here and that causes problems later on). We believe that this will turn up significant errors then in MGI (and Xenbase) noctua output that then get dropped.

Once we are able to get a better lens on the issue by making sure that it is reporting, the second step is to determine where the "fix" should be made. That will either be in adjusting the upstream ECO mapping file (https://github.com/evidenceontology/evidenceontology/blob/master/gaf-eco-mapping.txt) or creating some special cases in the code.

**Comment by ukemi** [comment]
This might be the case for a lot of annotations, but not the ones that are missing. One of our load procedures strips out annotations that can't be mapped to a three-letter code as well as other reasons. Those are some of the MANY annotations that get stripped out of the file when we do our loads. However, annotations that are in our gaf made it into MGI, so they passed our checks. Those would be the ones that are missing. 
Here are a few of the lines that Lori filters out during our load procedure:
Invalid ECO id : cannot find valid GO Evidence Code : ECO:0005593
MGI	MGI:88276	is_active_in	GO:0098831	PMID:8909549	ECO:0005593			20180711	SynGO	part_of(UBERON:0002037)	contributor=https://orcid.org/0000-0003-3185-5709|noctua-model-id=gomodel:SYNGO_1000|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|contributor=https://orcid.org/0000-0002-0755-5899
****
Invalid ECO id : cannot find valid GO Evidence Code : ECO:0005593
MGI	MGI:88276	is_active_in	GO:0098831	PMID:8909549	ECO:0005593			20180711	SynGO	part_of(UBERON:0000956)	contributor=https://orcid.org/0000-0003-3185-5709|noctua-model-id=gomodel:SYNGO_1000|model-state=production|contributor=https://orcid.org/0000-0001-7351-8706|contributor=https://orcid.org/0000-0003-0938-8534|contributor=https://orcid.org/0000-0002-0755-5899
****
uberon id has > 1 emapa : UBERON:0001950	['EMAPA:32842', 'EMAPA:35589']
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0006062			20180711	SynGO	part_of(UBERON:0001950)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002
****
Invalid ECO id : cannot find valid GO Evidence Code : ECO:0005593
MGI	MGI:97314	is_active_in	GO:0099182	PMID:25869803	ECO:0005593			20180711	SynGO	part_of(UBERON:0001950)	contributor=https://orcid.org/0000-0001-5259-4945|model-state=production|noctua-model-id=gomodel:SYNGO_1002

Here are a few lines from Lori's load file that make it through our QC procedure:
GO:0098831	MGI:88276	J:36476	IDA			NOCTUA_SynGO	20180711		MGI	part_of&=&EMAPA:17787&==&contributor&=&https://orcid.org/0000-0003-3185-5709&==&noctua-model-id&=&gomodel:SYNGO_1000&==&model-state&=&production&==&contributor&=&https://orcid.org/0000-0001-7351-8706&==&contributor&=&https://orcid.org/0000-0003-0938-8534&==&contributor&=&https://orcid.org/0000-0002-0755-5899&==&go_qualifier&=&is_active_in&==&evidence&=&ECO:0006003
GO:0098831	MGI:88276	J:36476	IDA			NOCTUA_SynGO	20180711		MGI	part_of&=&EMAPA:17544&==&contributor&=&https://orcid.org/0000-0003-3185-5709&==&noctua-model-id&=&gomodel:SYNGO_1000&==&model-state&=&production&==&contributor&=&https://orcid.org/0000-0001-7351-8706&==&contributor&=&https://orcid.org/0000-0003-0938-8534&==&contributor&=&https://orcid.org/0000-0002-0755-5899&==&go_qualifier&=&is_active_in&==&evidence&=&ECO:0006003
GO:0098685	MGI:88276	J:86623	IMP			NOCTUA_SynGO	20180711		MGI	contributor&=&https://orcid.org/0000-0003-3185-5709&==&model-state&=&production&==&contributor&=&https://orcid.org/0000-0001-7351-8706&==&contributor&=&https://orcid.org/0000-0003-0938-8534&==&noctua-model-id&=&gomodel:SYNGO_1001&==&contributor&=&https://orcid.org/0000-0002-0755-5899&==&go_qualifier&=&is_active_in&==&evidence&=&ECO:0001225

**Comment by kltm** [comment]
@ukemi To give a little more background on what's happening here. Within the pipeline, there is a step where it goes, essentially TTL -> (GPAD-ish model) -> GAF. This step was, for some cases, able to emit illegal GAF related to ECO mapping; this illegal GAF was then getting caught in another bit of QC and silently disappearing. This meant that essentially we'd get a silent reduction in annotations late in the pipeline with little associated record, but still be producing perfectly fine GAFs and GPADs otherwise. This seems to likely be the cause of the issue here with the SynGO reduction (testing), and possibly some other tickets we have.

**Comment by ukemi** [comment]
Yes. I understand, but I think those were also the annotations getting filtered out in our load pipeline as above. There are still ones that DO map to valid 3-letter codes and generate legal gaf lines. I think those are the ones that have gone missing. So unless the step was throwing out both the legal lines and the illegal lines, then I still don't understand where the annotations went that were loaded into MGI.

**Comment by kltm** [comment]
@dustine32 After much banging of head against wall on https://github.com/geneontology/pipeline/issues/284, I'm now testing your go-site `master` patch on pipeline `snapshot`.

**Comment by kltm** [comment]
We now have some product we can view for this ticket. As expected, the problem is the same:
```
$ zgrep -i syngo mgi.gaf.gz  | wc -l
312
```
Unfortunately, we did not see the expected uptick in errors post patch at http://skyhook.berkeleybop.org/snapshot/reports/noctua_mgi-report.html#gorule-0000001. Hm.
```
$ cat requirements.txt 
PyYAML
ontobio==2.8.0
click
pyparsing==2.4.7
antlr4-python3-runtime==4.9.3
```
Looks right.

@dustine32 , am I misremembering that what you did should already be propagated?

**Comment by dustine32** [comment]
@kltm Correct. https://github.com/biolink/ontobio/pull/621 was merged and should be in `2.8.0`. I was also expecting [noctua_mgi_valid.gaf.gz](http://snapshot.geneontology.org/products/annotations/noctua_mgi_valid.gaf.gz) to start dropping most of the SynGO lines but I'm still seeing the usual 6919 count. Trying to reproduce with a local `make target/groups/mgi/mgi.group` right now.

**Comment by dustine32** [comment]
Hm, the local `make mgi.group` run did filter the SynGO as designed. Gonna have to debug the Jenkins run further. Either the newer `2.8.0` ontobio isn't actually being run (I saw it did at least pip install `2.8.0`) or these SynGO lines in the Jenkins context aren't meeting the [conditions in the new check](https://github.com/biolink/ontobio/blob/0212592620c63a4f691907869aa50d0b50035830/ontobio/io/gpadparser.py#L207-L208).

**Comment by dustine32** [comment]
Looks like the new ECO check is actually being "skipped" in the later temp post filter step (Makefile-gaf-reprocess noctua_gpad) rather than the megamake step. This cmd below runs and the produced `noctua_mgi.gpad` and `noctua_mgi.report.md` files are then copied/overwritten to skyhook:
```
ontobio-parse-assocs.py -f /opt/go-site/noctua_sources/noctua_mgi-src.gpad -F gpad -o /opt/go-site/noctua_target/noctua_mgi.gpad -r /opt/go-site/go-ontology.json --report-md /opt/go-site/noctua_target/noctua_mgi.report.md --report-json /opt/go-site/noctua_target/noctua_mgi.report.json -l all convert --to gpad -n 1.2
```
Debugging this, I see that our new `--allow_unmapped_eco` flag is actually defaulted to `None` in `ontobio-parse-assocs.py` and this does not satisfy the new ECO check's condition `self.config.allow_unmapped_eco is False`. I will quickly fix this in ontobio and do another release.

**Comment by kltm** [comment]
Thank you @dustine32 for catching me up. Starting a test on `master`.

**Comment by kltm** [comment]
@dustine32 I wanted to confirm that the rule seems to now be working as intended: http://skyhook.berkeleybop.org/master/reports/noctua_mgi-report.html#gorule-0000001
I think that looks like the kind of catch that we'd be expecting from MGI?

If so, that violation list would be a place to work with @pgaudet (from the other thread) to figure out our next steps. I believe that we'd probably want to confirm the correct mapping from ECO to 3-letter codes gives enough coverage and is correct before looking at other measures.

**Comment by dustine32** [comment]
@kltm Correct. Looks like it's finally reporting those errors correctly.

**Comment by kltm** [comment]
It looks like https://github.com/geneontology/pipeline/issues/283 has had a major effect:
```
$ zgrep -i syngo ./products/annotations/mgi-src.gaf.gz | wc -l
155
$ zgrep -i syngo ./products/annotations/noctua_mgi_valid.gaf.gz | wc -l
5093
$ zgrep -i syngo ./annotations/mgi.gaf.gz | wc -l 
5248
```
Is this bringing the number up to about where we'd expect?

**Comment by malcolmfisher103** [comment]
The error reporting for Xenbase looks right (http://skyhook.berkeleybop.org/master/reports/noctua_xenbase-report.html#gorule-0000001), with ~450 mostly morpholino based annotations being flagged as errors.

**Comment by dustine32** [comment]
@kltm That's actually a bit higher than expected when comparing with the last (2022-03-22) [release mgi.gaf](//release.geneontology.org/2022-03-22/annotations/mgi.gaf.gz):
```
$ grep -i syngo mgi.gaf | wc -l
    4672
```
Assuming the upstream SynGO data in Noctua hasn't changed yet.

I think this jump is OK as MGI's pipeline was previously kicking out SynGO annotations for things that we don't check for as @ukemi [mentioned](https://github.com/geneontology/go-site/issues/1847#issuecomment-1117222983) above. For example:

> uberon id has > 1 emapa : UBERON:0001950 ['EMAPA:32842', 'EMAPA:35589']

Wonder if the difference `5248 - 4672 ==` # of UBERON->EMAPA cardinality errors MGI's pipeline reported?

**Comment by ukemi** [comment]
Sorry guys. I was a bit misleading with the above. In the cases where the annotation extension is a problem, the annotation still gets loaded but the annotation extension gets stripped because the 'core' annotation is still valid.

**Comment by kltm** [comment]
Looking at
```
taxon_subset_closure_label: Mus musculus
```
between the last `release` and current `snapshot`, mouse annotations look like:

release:
(272347) | MGI
(3309) | SynGO

snapshot:
(301146) | MGI
(3592) | SynGO
